### ISSUE TITLE
max_list/2 and min_list/2 problem: If Max or Min are bound on-call you must resort to comparing numerically at the end

### ISSUE THREAD

**Issue by dtonhofer** [SWI-Prolog/swipl-devel#808]
With the current implementation:

```
?- max_list([1.0 , 1],N).
N = 1.

?- max_list([1 , 1.0],N).
N = 1.0.
```

The above has unfortunate effects:

```
?- max_list([1.0 , 1] , 1.0).
false.

?- max_list([1.0 , 1] , 1).
true.
```

😓

... which can only be fixed by "grabbing and comparing numerically":

```
?- max_list([1.0 , 1] , M), M =:= 1.0.
M = 1.

?- max_list([1 , 1.0] , M), M =:= 1.0.
M = 1.0.
```

Thus the code should be:

```
max_list([H|T], Max) :-
   max_list(T, H, Max1),  % find it
   (var(Max) -> (Max = Max1) 
    ;
    number(Max) -> (Max =:= Max1) % perform type conversion if needed
    ;
   fail). % currently fails on non-number

max_list([], Max0, Max) =>   % (Does it need to be written like that or can one just unify in the head?)
  Max = Max0.

max_list([H|T], Max0, Max) =>
  Max1 is max(H, Max0), % Should maybe decouple from max/2 and just use ">" arithmetic comparison
  max_list(T, Max1, Max).
```

**P.S.**

Maybe there should be a max_list_all/2 which collects all the maxes as they appear

```
max_list_all([], []) :- !.

max_list_all([X|Xs], Result) :-
   max_list_all_2(Xs, X, [X|Fin], Fin, Result).   

% max_list_all_2(Xs, MaxSoFar, OpenListOfMax, OpenListOfMaxFin, Result).

max_list_all_2([], _, Result, [], Result). % at the end, close the open list which is the Result

max_list_all_2([X|Xs], MaxSoFar, OpenListOfMax, OpenListOfMaxFin, Result) :-
   X < MaxSoFar, !,
   max_list_all_2(Xs, MaxSoFar, OpenListOfMax, OpenListOfMaxFin, Result).
   
max_list_all_2([X|Xs], MaxSoFar, _, _, Result) :-
   X > MaxSoFar, !,
   max_list_all_2(Xs, X, [X|Fin], Fin, Result).

max_list_all_2([X|Xs], MaxSoFar, OpenListOfMax, OpenListOfMaxFin, Result) :-
   X =:= MaxSoFar, !,
   OpenListOfMaxFin = [X|NewFin],
   max_list_all_2(Xs, MaxSoFar, OpenListOfMax, NewFin, Result).
   

:- begin_tests(max_list_all).

test(1) :- 
   max_list_all([],[]).
   
test(2) :- 
   max_list_all([1,2,3],R),
   assertion(R == [3]).
   
test(3) :- 
   max_list_all([3,2,1],R),
   assertion(R == [3]).
   
test(4) :- 
   max_list_all([3,2,3],R),
   assertion(R == [3,3]).
   
test(5) :- 
   max_list_all([3.0,2,3],R),
   assertion(R == [3.0,3]).
   
test(6) :- 
   max_list_all([3,2,3.0],R),
   assertion(R == [3,3.0]).
   
test(7) :- 
   max_list_all([1,2,3,4,4,4],R),
   assertion(R == [4,4,4]).
   
test(8) :- 
   max_list_all([4,4,4,3,2,1],R),
   assertion(R == [4,4,4]).
   
test(9) :- 
   max_list_all([1,2,4,4,4,2,1],R),
   assertion(R == [4,4,4]).
   
:- end_tests(max_list_all).
```

**Comment by JanWielemaker** [comment]
If anything, I'd say that the _function_ max/2 should not pick an arbitrary member if both are equal.  Max is defined to return the max of a set of floating point numbers as a float by ISO.  SWI-Prolog doesn't follow that and returns the largest number without converting.    If the two are equal it should probably return the one of the type used for the comparison.

**Comment by JanWielemaker** [comment]
Addressed in cb55f2c3da164d61c5f684482653bee1c31d51f3
